torch_geometric/explain/glg_explainer.py

```diff
@@ -0,0 +1,621 @@
+
+    For an example of using :class:`GLGExplainer`,
+        see `examples/glg_explainer.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/examples/
+        /glg_explainer.py>`_.
+
+    Args:
+        le_embedder (torch.nn.Module): The Local Explanations Embedder.
+            It must be a :class:`torch.nn.Module` providing an embedding
+            for each input local explanation
+        num_classes (int): Number of output classes
+        num_prototypes (int): Number of prototypes to use
+        dim_prototypes (int, optional): Latent dimensionality of each
+            prototype. (default :obj:`10`)
+        dim_len_hidden (int, optional): Latent dimensionality of the E-LEN.
+            (default :obj:`10`)
+        len_lr (float, optional): Learning rate for the parameters of the
+            E-LEN. (default :obj:`0.0005`)
+        prototypes_lr (float, optional): Learning rate for the prototypes.
+            (default :obj:`0.001`)
+        le_embedder_lr (float, optional): Learning rate for the parameters
+            of the :attr:`le_embedder`. (default :obj:`0.001`)
+        coeff_r1 (float, optional): Scaling weight for the distance loss
+            :obj:`R1`. (default :obj:`0.09`)
+        coeff_r2 (float, optional): Scaling weight for the distance loss
+            :obj:`R2`. (default :obj:`0.00099`)
+    """
+    def __init__(
+        self,
+        le_embedder: torch.nn.Module,
+        num_classes: int,
+        num_prototypes: int,
+        dim_prototypes: Optional[int] = 10,
+        dim_len_hidden: Optional[int] = 10,
+        len_lr: Optional[float] = 0.0005,
+        prototypes_lr: Optional[float] = 0.001,
+        le_embedder_lr: Optional[float] = 0.001,
+        coeff_r1: Optional[float] = 0.09,
+        coeff_r2: Optional[float] = 0.00099,
+    ):
+        super().__init__()
+
+        self.le_embedder = le_embedder
+        self.num_classes = num_classes
+        self.coeff_r1 = coeff_r1
+        self.coeff_r2 = coeff_r2
+        self.global_explanation = LogicExplanations()
+        self.len_loss = self._resolve_len_loss()
+
+        self.len_model = _ELEN(
+            input_shape=num_prototypes,
+            hidden_dim=dim_len_hidden,
+            n_classes=num_classes
+        )
+        self.prototype_vectors = torch.nn.Parameter(
+            torch.rand((num_prototypes,
+                        dim_prototypes)),
+            requires_grad=True
+        )
+        self.optimizer = torch.optim.Adam(
+            self.le_embedder.parameters(),
+            lr=le_embedder_lr
+        )
+        self.optimizer.add_param_group({
+            'params': self.len_model.parameters(),
+            'lr': len_lr
+        })
+        self.optimizer.add_param_group({
+            'params': self.prototype_vectors,
+            'lr': prototypes_lr
+        })
+
+    def do_epoch(self, loader: DataLoader, train: bool = False) -> Dict:
```

Comment:
Can we create a `train` interface like [PGExplainer](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/explain/algorithm/pg_explainer.py), where its expects inputs `x`, `edge_index` and `batch` and runs once batch of training.